tests/specs/lemmas.k

```diff
@@ -201,6 +205,40 @@ module LEMMAS
 
     rule chop(I) => I requires 0 <=Int I andBool I <Int pow256  [simplification]
 
+    rule N <=Int maxInt(P, Q) => true requires N <=Int P  orBool N <=Int Q  [simplification]
+    rule N  <Int maxInt(P, Q) => true requires N  <Int P  orBool N  <Int Q  [simplification]
+    rule N >=Int maxInt(P, Q) => true requires N >=Int P andBool N >=Int Q  [simplification]
+    rule N  >Int maxInt(P, Q) => true requires N  >Int P andBool N  >Int Q  [simplification]
+    rule maxInt(P, Q) >=Int N => true requires N <=Int P  orBool N <=Int Q  [simplification]
+    rule maxInt(P, Q)  >Int N => true requires N  <Int P  orBool N  <Int Q  [simplification]
+    rule maxInt(P, Q) <=Int N => true requires N >=Int P andBool N >=Int Q  [simplification]
+    rule maxInt(P, Q)  <Int N => true requires N  >Int P andBool N  >Int Q  [simplification]
+
+    rule N -Int maxInt(P, Q) => minInt(N -Int P, N -Int Q)  [simplification]
+    rule N +Int maxInt(P, Q) => maxInt(N +Int P, N +Int Q)  [simplification]
+    rule maxInt(P, Q) -Int N => maxInt(P -Int N, Q -Int N)  [simplification]
+    rule maxInt(P, Q) +Int N => maxInt(P +Int N, Q +Int N)  [simplification]
+
+    rule N <=Int minInt(P, Q) => true requires N <=Int P andBool N <=Int Q  [simplification]
+    rule N  <Int minInt(P, Q) => true requires N  <Int P andBool N  <Int Q  [simplification]
+    rule N >=Int minInt(P, Q) => true requires N >=Int P  orBool N >=Int Q  [simplification]
+    rule N  >Int minInt(P, Q) => true requires N  >Int P  orBool N  >Int Q  [simplification]
+    rule minInt(P, Q) >=Int N => true requires N <=Int P andBool N <=Int Q  [simplification]
+    rule minInt(P, Q)  >Int N => true requires N  <Int P andBool N  <Int Q  [simplification]
+    rule minInt(P, Q) <=Int N => true requires N >=Int P  orBool N >=Int Q  [simplification]
+    rule minInt(P, Q)  <Int N => true requires N  >Int P  orBool N  >Int Q  [simplification]
+
+    rule N -Int minInt(P, Q) => maxInt(N -Int P, N -Int Q)  [simplification]
+    rule N +Int minInt(P, Q) => minInt(N +Int P, N +Int Q)  [simplification]
+    rule minInt(P, Q) -Int N => minInt(P -Int N, Q -Int N)  [simplification]
+    rule minInt(P, Q) +Int N => minInt(P +Int N, Q +Int N)  [simplification]
+
+    rule N >=Int #sizeByteArray(#asByteStack(P)) => true  requires 1 <<Int (N <<Int 3)  >Int P andBool N >Int 0 [simplification]  // Do I need N >Int 0?
+    rule N  <Int #sizeByteArray(#asByteStack(P)) => true  requires 1 <<Int (N <<Int 3) <=Int P                  [simplification]
+    rule N  <Int #sizeByteArray(#asByteStack(P)) => false requires 1 <<Int (N <<Int 3)  >Int P andBool N >Int 0 [simplification]  // Do I need N >Int 0?
```

Comment:
These lemmas might seriously affect performance. Do you really need them? Can haskell prove the requires clause for them using `<<Int` operator? Multiplication might work better, at least instead of `N <<Int 3`

Reply:
These were needed in cases where `N` was concrete, where everything worked fine. We can, in principle, send `<<Int` to the SMT solver, but I don't actually know what would have happened if `N` was symbolic.

This PR contains only a part of what the Haskell backend would need anyway, so I can remove these and leave them for a later PR.